numpy/core/src/umath/ufunc_object.c

```diff
@@ -2005,9 +2176,28 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
         retval = -1;
         goto fail;
     }
+    
+    /* Possibly remap axes. */
+    if (axes) {
+        remap_axis = PyArray_malloc(sizeof(remap_axis[0]) * nop);
+        remap_axis_memory = PyArray_malloc(sizeof(remap_axis_memory[0]) *
+                                                  nop * NPY_MAXDIMS);
```

Comment:
It would be nice to avoid mallocs. In principle we could replace `nop` by `NPY_MAXARGS` and allocate on the stack, but that would be `32*32` ints, or 8kb (4kb on 32bit), which seems a little large to put on the stack. But I see other places in numpy we have already done this, eg a grep shows:
```
numpy/core/src/umath/ufunc_object.c:2061:    int op_axes_arrays[NPY_MAXARGS][NPY_MAXDIMS];
numpy/core/src/multiarray/nditer_pywrap.c:742:    int op_axes_arrays[NPY_MAXARGS][NPY_MAXDIMS];
numpy/core/src/multiarray/einsum.c.src:2610:    char op_labels[NPY_MAXARGS][NPY_MAXDIMS];
numpy/core/src/multiarray/einsum.c.src:2617:    int op_axes_arrays[NPY_MAXARGS][NPY_MAXDIMS];
```

I like the strategy in the 3rd line there to use a `char`, so it is only 1kb. I also like how those other examples use C's 2d static array syntax, instead of allocating a separate index array.

In summary, I actually think it would be a significant improvement to allocate `remap_axis_memory` as a static 2d char array.

Reply:
I had that originally, but thought that it was a waste of memory for what should normally be a rare use case. A yet more complicated alternative would be to allocate a, say, 3x3 2-D array and not use it is there are more arguments. (note that the `char` is for op_labels, which are in fact characters.)

Reply:
There are a few ways to tidy this along those lines. 

Perhaps do this: Get rid of `remap_axis` and only have  `char remap_axis_memory[NPY_MAXARGS][NPY_MAXDIMS];` Then redefine `#define REMAP_AXIS(iop, axis) ((axes ? remap_axis_memory[iop][axis] : axis)`

Reply:
Oh sorry, I didn't see your comment.  Yeah it is debatable.

I guess if you already investigated both options, I am happy to go with whichever you thin kis better.

Reply:
Any reason not to combine these into one allocation?

Reply:
`remap_axis` is of pointers, `remap_axis_memory` of `int`, so in principle they can have different stride size, correct?

Reply:
Stride size isn't relevant if you're allocating a `void*`. I suppose alignment could be an issue?

Reply:
In the end, I think this is an implementation detail - the present code is at least somewhat clear, so maybe best to leave it for later.